subprojects/base-services/src/main/java/org/gradle/internal/service/MethodHandleBasedServiceMethodFactory.java

```diff
@@ -16,14 +16,18 @@
 package org.gradle.internal.service;
 
 import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
 
 class MethodHandleBasedServiceMethodFactory implements ServiceMethodFactory {
     @Override
     public ServiceMethod toServiceMethod(Method method) {
-        try {
-            return new MethodHandleBasedServiceMethod(method);
-        } catch (IllegalAccessException ex) {
-            return new ReflectionBasedServiceMethod(method);
+        if (Modifier.isPublic(method.getModifiers()) && Modifier.isPublic(method.getDeclaringClass().getModifiers())) {
```

Comment:
Yes, and it is a problem. It won't work in JDK 9 anymore, because `setAccessible(true)` won't be allowed, like we do for `JavaMethod`, for example.